File: OrderController.js
/**
 * McDonald's Order Controller
 * Manages orders and cooking bots for automated order processing
 */

export class OrderController {
  constructor() {
    this.orders = [];
    this.bots = [];
    this.completedOrders = [];
    this.orderCounter = 1000;
    this.botCounter = 0;
  }

  /**
   * Create a new order
   * @param {string} type - Order type: 'NORMAL' or 'VIP'
   * @returns {object} Created order
   */
  createOrder(type = 'NORMAL') {
    this.orderCounter++;
    const order = {
      id: this.orderCounter,
      type: type.toUpperCase(),
      status: 'PENDING',
      createdAt: new Date(),
    };

    // VIP orders go to the front, but behind other VIP orders
    if (type.toUpperCase() === 'VIP') {
      const lastVipIndex = this.orders.findLastIndex(o => o.type === 'VIP');
      this.orders.splice(lastVipIndex + 1, 0, order);
    } else {
      this.orders.push(order);
    }

    return order;
  }

  /**
   * Add a new bot
   * @returns {object} Created bot
   */
  addBot() {
    this.botCounter++;
    const bot = {
      id: this.botCounter,
      status: 'IDLE',
      currentOrder: null,
      processingTimer: null,
    };

    this.bots.push(bot);
    this.assignOrderToBot(bot);

    return bot;
  }

  /**
   * Remove a bot (removes the newest bot)
   * @returns {object|null} Removed bot or null if no bots
   */
  removeBot() {
    if (this.bots.length === 0) return null;

    const bot = this.bots[this.bots.length - 1];

    // If bot is processing an order, return it to pending
    if (bot.status === 'PROCESSING' && bot.currentOrder) {
      if (bot.processingTimer) {
        clearTimeout(bot.processingTimer);
      }
      
      const order = bot.currentOrder;
      order.status = 'PENDING';
      this.orders.unshift(order); // Put at front of queue
    }

    this.bots.pop();
    return bot;
  }

  /**
   * Assign a pending order to an idle bot
   * @param {object} bot - The bot to assign work to
   */
  assignOrderToBot(bot) {
    if (this.orders.length === 0) {
      bot.status = 'IDLE';
      return;
    }

    const order = this.orders.shift();
    bot.currentOrder = order;
    bot.status = 'PROCESSING';
    order.status = 'PROCESSING';

    // Simulate 10-second processing time
    bot.processingTimer = setTimeout(() => {
      this.completeOrder(bot, order);
    }, 10000);
  }

  /**
   * Complete an order
   * @param {object} bot - Bot that completed the order
   * @param {object} order - Order that was completed
   */
  completeOrder(bot, order) {
    order.status = 'COMPLETE';
    order.completedAt = new Date();
    this.completedOrders.push(order);

    bot.currentOrder = null;
    bot.processingTimer = null;

    // Try to assign next order
    this.assignOrderToBot(bot);
  }

  /**
   * Get all pending orders
   * @returns {array} Array of pending orders
   */
  getPendingOrders() {
    return this.orders;
  }

  /**
   * Get all completed orders
   * @returns {array} Array of completed orders
   */
  getCompletedOrders() {
    return this.completedOrders;
  }

  /**
   * Get all bots
   * @returns {array} Array of bots
   */
  getBots() {
    return this.bots;
  }

  /**
   * Get system status
   * @returns {object} Current system status
   */
  getStatus() {
    return {
      totalBots: this.bots.length,
      activeBots: this.bots.filter(b => b.status === 'PROCESSING').length,
      idleBots: this.bots.filter(b => b.status === 'IDLE').length,
      pendingOrders: this.orders.length,
      completedOrders: this.completedOrders.length,
    };
  }
}

File: order_controller.go
package main

import (
	"fmt"
	"sync"
	"time"
)

type OrderType int

const (
	NormalOrder OrderType = iota
	VIPOrder
)

type OrderStatus int

const (
	Pending OrderStatus = iota
	Processing
	Complete
)

type Order struct {
	ID     int
	Type   OrderType
	Status OrderStatus
}

func (o *Order) String() string {
	orderType := "Normal"
	if o.Type == VIPOrder {
		orderType = "VIP"
	}
	return fmt.Sprintf("%s#%d", orderType, o.ID)
}

type Bot struct {
	ID            int
	CurrentOrder  *Order
	IsProcessing  bool
	stopChan      chan struct{}
	completedChan chan *Order
}

type OrderController struct {
	mu          sync.Mutex
	orders      []*Order
	nextOrderID int
	bots        []*Bot
	nextBotID   int
	pendingQ    []*Order
}

func NewOrderController() *OrderController {
	return &OrderController{
		orders:      make([]*Order, 0),
		nextOrderID: 1,
		bots:        make([]*Bot, 0),
		nextBotID:   1,
		pendingQ:    make([]*Order, 0),
	}
}

func (oc *OrderController) AddNormalOrder() *Order {
	oc.mu.Lock()
	defer oc.mu.Unlock()

	order := &Order{
		ID:     oc.nextOrderID,
		Type:   NormalOrder,
		Status: Pending,
	}
	oc.nextOrderID++
	oc.orders = append(oc.orders, order)
	oc.insertIntoPending(order)

	fmt.Printf("[%s] Added %s\n", time.Now().Format("15:04:05"), order)
	oc.printStatusNoLock()

	return order
}

func (oc *OrderController) AddVIPOrder() *Order {
	oc.mu.Lock()
	defer oc.mu.Unlock()

	order := &Order{
		ID:     oc.nextOrderID,
		Type:   VIPOrder,
		Status: Pending,
	}
	oc.nextOrderID++
	oc.orders = append(oc.orders, order)
	oc.insertIntoPending(order)

	fmt.Printf("[%s] Added %s\n", time.Now().Format("15:04:05"), order)
	oc.printStatusNoLock()

	return order
}

func (oc *OrderController) insertIntoPending(order *Order) {
	insertIdx := len(oc.pendingQ)
	for i, o := range oc.pendingQ {
		if o.Type == NormalOrder {
			insertIdx = i
			break
		}
	}

	if insertIdx == len(oc.pendingQ) {
		oc.pendingQ = append(oc.pendingQ, order)
	} else {
		oc.pendingQ = append(oc.pendingQ[:insertIdx], append([]*Order{order}, oc.pendingQ[insertIdx:]...)...)
	}
}

func (oc *OrderController) AddBot() *Bot {
	oc.mu.Lock()
	defer oc.mu.Unlock()

	bot := &Bot{
		ID:            oc.nextBotID,
		IsProcessing:  false,
		stopChan:      make(chan struct{}),
		completedChan: make(chan *Order, 1),
	}
	oc.nextBotID++
	oc.bots = append(oc.bots, bot)

	fmt.Printf("[%s] Added bot #%d\n", time.Now().Format("15:04:05"), bot.ID)
	oc.printStatusNoLock()

	oc.tryAssignOrderAsync(bot)

	return bot
}

func (oc *OrderController) RemoveBot() {
	oc.mu.Lock()
	defer oc.mu.Unlock()

	if len(oc.bots) == 0 {
		fmt.Printf("[%s] No bots to remove\n", time.Now().Format("15:04:05"))
		return
	}

	lastBot := oc.bots[len(oc.bots)-1]
	oc.bots = oc.bots[:len(oc.bots)-1]

	if lastBot.IsProcessing && lastBot.CurrentOrder != nil {
		close(lastBot.stopChan)

		order := lastBot.CurrentOrder
		order.Status = Pending

		lastBot.IsProcessing = false
		lastBot.CurrentOrder = nil

		oc.insertIntoPending(order)
		fmt.Printf("[%s] Removed bot #%d (was processing %s, returned to pending)\n",
			time.Now().Format("15:04:05"), lastBot.ID, order)
	} else {
		fmt.Printf("[%s] Removed bot #%d (idle)\n", time.Now().Format("15:04:05"), lastBot.ID)
	}

	oc.printStatusNoLock()
}

func (oc *OrderController) tryAssignOrderAsync(bot *Bot) {
	go func() {
		for {
			var order *Order

			oc.mu.Lock()

			if len(oc.pendingQ) > 0 {
				order = oc.pendingQ[0]
				oc.pendingQ = oc.pendingQ[1:]

				order.Status = Processing
				bot.IsProcessing = true
				bot.CurrentOrder = order
			}

			oc.mu.Unlock()

			if order == nil {
				time.Sleep(100 * time.Millisecond)
				continue
			}

			fmt.Printf("[%s] Bot #%d started processing %s\n",
				time.Now().Format("15:04:05"), bot.ID, order)
			oc.printStatus()

			select {
			case <-time.After(10 * time.Second):
				oc.mu.Lock()
				order.Status = Complete
				bot.IsProcessing = false
				bot.CurrentOrder = nil
				oc.mu.Unlock()

				fmt.Printf("[%s] Bot #%d completed %s\n",
					time.Now().Format("15:04:05"), bot.ID, order)
				oc.printStatus()

			case <-bot.stopChan:
				fmt.Printf("[%s] Bot #%d stopped while processing %s\n",
					time.Now().Format("15:04:05"), bot.ID, order)
				return
			}
		}
	}()
}

func (oc *OrderController) printStatus() {
	oc.mu.Lock()
	defer oc.mu.Unlock()
	oc.printStatusNoLock()
}

func (oc *OrderController) printStatusNoLock() {
	botCount := len(oc.bots)

	var pendingOrders []string
	for _, order := range oc.pendingQ {
		pendingOrders = append(pendingOrders, order.String())
	}

	var processingOrders []string
	for _, bot := range oc.bots {
		if bot.IsProcessing && bot.CurrentOrder != nil {
			processingOrders = append(processingOrders, bot.CurrentOrder.String())
		}
	}

	allOrders := append(pendingOrders, processingOrders...)

	if len(allOrders) == 0 {
		fmt.Printf("[%s] status: bot: [%d], order: []\n",
			time.Now().Format("15:04:05"), botCount)
	} else {
		fmt.Printf("[%s] status: bot: [%d], order: [%s]\n",
			time.Now().Format("15:04:05"), botCount, joinStrings(allOrders))
	}
}

func joinStrings(strs []string) string {
	if len(strs) == 0 {
		return ""
	}
	result := strs[0]
	for i := 1; i < len(strs); i++ {
		result += ", " + strs[i]
	}
	return result
}

func (oc *OrderController) GetPendingCount() int {
	oc.mu.Lock()
	defer oc.mu.Unlock()
	return len(oc.pendingQ)
}

func (oc *OrderController) GetBotCount() int {
	oc.mu.Lock()
	defer oc.mu.Unlock()
	return len(oc.bots)
}

func (oc *OrderController) GetCompletedCount() int {
	oc.mu.Lock()
	defer oc.mu.Unlock()
	count := 0
	for _, order := range oc.orders {
		if order.Status == Complete {
			count++
		}
	}
	return count
}
